File: 0907-koko-eating-bananas/0907-koko-eating-bananas.cpp
#include <cmath>
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int left = 1, right = *max_element(piles.begin(), piles.end());
        
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (canFinish(piles, mid, h)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        
        return left;
    }
    

     bool canFinish(vector<int>& piles, int k, int h) {
        int hours = 0;
        for (int pile : piles) {
            hours += (pile + k - 1) / k;  // Using this to compute ceiling of division
        }
        return hours <= h;
    }
};

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int left = 1, right = 1;
        for(int pile : piles){
            right = Math.max ( right, pile);
        }
        while(left < right){
            int mid = left + ( right - left ) / 2;
            if(canFinish(piles, mid, h)){
                right = mid;
            }
            else{
                left = mid+1;
            }
        }
        return left;
    }
    public static boolean canFinish(int[] piles, int k, int h){
        int hours = 0;
        for(int pile : piles){
            hours += Math.ceil((double)pile/k);
        }
        return hours <= h;
    }
}